iceoryx2-bb/container/src/flatmap.rs

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) 2025 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache Software License 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
+// which is available at https://opensource.org/licenses/MIT.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! A FlatMap is a data structure to store key-value pairs. The
+//! [`FixedSizeFlatMap`](crate::flatmap::FixedSizeFlatMap) is a compile-time fixed-size flatmap
+//! that is self-contained and shared-memory compatible.
+//!
+//! # User Examples
+//!
+//! ```
+//! use iceoryx2_bb_container::flatmap::FixedSizeFlatMap;
+//!
+//! const CAPACITY: usize = 100;
+//! let mut map = FixedSizeFlatMap::<u8, u8, CAPACITY>::new();
+//! assert_eq!(map.insert(23, 4).is_ok(), true);
+//! assert_eq!(map.get(&23).unwrap(), 4);
+//! ```
+
+use crate::slotmap::FixedSizeSlotMap;
+use iceoryx2_bb_elementary_traits::{
+    placement_default::PlacementDefault, zero_copy_send::ZeroCopySend,
+};
+use iceoryx2_bb_log::fail;
+
+/// Failures caused by [`FixedSizeFlatMap::insert()`]
+#[derive(Debug, Clone, Copy, Eq, Hash, PartialEq)]
+pub enum FlatMapError {
+    /// The FlatMap already contains the key that shall be inserted.
+    KeyAlreadyExists,
+    /// The FlatMap is full and cannot hold an additional key-value pair.
+    IsFull,
+}
+
+#[repr(C)]
+struct Entry<K: Eq, V: Clone> {
+    id: K,
+    value: V,
+}
+
+unsafe impl<K: Eq + ZeroCopySend, V: Clone + ZeroCopySend> ZeroCopySend for Entry<K, V> {}
+
+/// A data structure to store key-value pairs
+#[repr(C)]
+pub struct FixedSizeFlatMap<K: Eq + ZeroCopySend, V: Clone + ZeroCopySend, const CAPACITY: usize> {
```

Comment:
I think we can get rid of the `+ ZeroCopySend` requirement for `K` and `V` in the struct definition since you implement `ZeroCopySend` for the `FixedSizeFlatmap` only when both, `K` and `V` implement `ZeroCopySend`, see line 57.

This would also mean that you remove `+ ZeroCopySend` from every impl except the `ZeroCopySend for FixedSizeFlatmap` line.